### ISSUE TITLE
TLS 1.3 server has no preference Error

### ISSUE THREAD

**Issue by S8Cloud** [hakasenyang/openssl-patch#9]
When using default openssl 1.1.1 without any patch and enable tls1.3 feature, the ssllabs test will show suites in server-preferred order in the cipher part.

However, if your equal patch or openssl prefer patch enabled, the tls 1.3 part will show server has no preference.

**Comment by hakasenyang** [comment]
It is not a bug.

You can modify this syntax as follows:
[TLS13+AESGCM+AES128|TLS13+AESGCM+AES256|TLS13+CHACHA20]
to
TLS13+AESGCM+AES128:TLS13+CHACHA20:TLS13+AESGCM+AES256

TLS 1.3 now has a total of five Cipher (including AESCCM).
We currently use three, and we are using Equal Preference for three, so the priority is in the browser.

The results of SSLLabs are correct, and tests can be tested using the **-cipher** option of **openssl s_client**.

### Example
If you have already completed the openssl 1.1.1 equal preference patch (except for the _ciphers patch)

#### Original (If you have not changed the above Cipher)
- openssl s_client -connect localhost:443 -cipher TLS13+AESGCM+AES128:TLS13+AESGCM+AES256<br>Result: AES128
- openssl s_client -connect localhost:443 -cipher TLS13+AESGCM+AES256:TLS13+AESGCM+AES128<br>Result: AES256

#### Changed
- openssl s_client -connect localhost:443 -cipher TLS13+AESGCM+AES128:TLS13+AESGCM+AES256<br>Result: AES128
- openssl s_client -connect localhost:443 -cipher TLS13+AESGCM+AES256:TLS13+AESGCM+AES128<br>Result: AES128

**Comment by S8Cloud** [comment]
![d0759438-8cb5-4b7a-815f-f5f1bf2cdd96](https://user-images.githubusercontent.com/27051990/47254115-07709d00-d490-11e8-83f3-b1c667866dd3.jpeg)

This is result of your blog 

![5cb21a4f-e2cf-4f4d-96a3-b45d6d895784](https://user-images.githubusercontent.com/27051990/47254116-07709d00-d490-11e8-928b-7adf26140702.jpeg)

And this comes from default Openssl 1.1.1

**Comment by hakasenyang** [comment]
Yes, I saw the data from Cloudflare and Google and applied it as is.

If you do not want to use it, please use the above settings.

**Comment by S8Cloud** [comment]
<img width="908" alt="2018-10-20 8 21 04" src="https://user-images.githubusercontent.com/27051990/47256647-02264900-d4b6-11e8-8758-ba7193dc6658.png">

Tried to set tls 1.3 cipher like `TLS13+AES128:TLS13+CHACHA20` with `openssl-equal-1.1.1_ciphers.patch` and the result is great 👍 

It seems the key is to make tls 1.3 cipher configurable by nginx (which is the main different of your 2 openssl patch ).

Is this feature based on the boringssl equal preference or can be separated and applied to openssl along?

**Comment by hakasenyang** [comment]
I'm sorry, I didn't quite understand what you wanted.

BoringSSL's equal preference patch only takes priority decisions. It is not related to the TLS 1.3 cipher setting.

In OpenSSL, several functions have been created for setting TLS 1.3 Cipher. So I simply put it back in a way that I could set up earlier.

If you want to configure TLS 1.3 Cipher in nginx, see the this patch diff. (Of course there are some more sources, but if you like, I will make a patch file for you.)



```diff
diff --git a/ssl/s3_lib.c b/ssl/s3_lib.c
index 7713f767b2..a0af8ac001 100644
--- a/ssl/s3_lib.c
+++ b/ssl/s3_lib.c
@@ -31,7 +31,25 @@ const unsigned char tls12downgrade[] = {
 };
 
 /* The list of available TLSv1.3 ciphers */
+/* Since nginx can not set the TLS 1.3 cipher, remove it temporarily. */
 static SSL_CIPHER tls13_ciphers[] = {
+    {
+        0,
+    }
+};
+
+/*
+ * The list of available ciphers, mostly organized into the following
+ * groups:
+ *      Always there
+ *      EC
+ *      PSK
+ *      SRP (within that: RSA EC PSK)
+ *      Cipher families: Chacha/poly, Camellia, Gost, IDEA, SEED
+ *      Weak ciphers
+ */
+static SSL_CIPHER ssl3_ciphers[] = {
+    /* TLSv1.3 ciphers */
     {
         1,
         TLS1_3_RFC_AES_128_GCM_SHA256,
@@ -111,20 +129,8 @@ static SSL_CIPHER tls13_ciphers[] = {
         SSL_HANDSHAKE_MAC_SHA256,
         128,
         128,
-    }
-};
-
-/*
- * The list of available ciphers, mostly organized into the following
- * groups:
- *      Always there
- *      EC
- *      PSK
- *      SRP (within that: RSA EC PSK)
- *      Cipher families: Chacha/poly, Camellia, Gost, IDEA, SEED
- *      Weak ciphers
- */
-static SSL_CIPHER ssl3_ciphers[] = {
+    },
+    /* List of cipher below TLSv1.3 */
     {
      1,
      SSL3_TXT_RSA_NULL_MD5,
```

**Comment by S8Cloud** [comment]
Seems this patch doesn't work for me. I patched the patch and set `ssl_ciphers TLS13+AESGCM+AES128:TLS13+CHACHA20;` but AES256 cipher still appeals on the test result.
Would appreciate it if a patch could be made :)

**Comment by hakasenyang** [comment]
I did not accidentally upload some sources.
Please confirm that you uploaded the patch file.

https://github.com/hakasenyang/openssl-patch/blob/master/openssl-1.1.1-tls13_nginx_config.patch
